redash/query_runner/big_query.py

```diff
@@ -305,19 +305,67 @@ def _list_datasets(self):
             for project_id in projects
         ])
 
-        # Submit the query to BigQuery.
-        resp = service.jobs().query(projectId=primary_project_id, body={
-            'useLegacySql':False,
-            'query': query
-        }).execute()
+        rows = self._get_results(query)
 
         # Format results as
         # [(project_id, dataset_id)]
-        return map(
+        data = map(
             lambda row: (row['f'][0]['v'], row['f'][1]['v']),
-            resp.get("rows", [])
+            rows
         )
 
+        return data
+
+    def _get_results(self, query):
+        service = self._get_bigquery_service()
+        primary_project_id = self._get_project_id()
+
+        # Run the query on BigQuery.
+        resp = service.jobs().query(projectId=primary_project_id, body={
+            'useLegacySql':False,
+            'query': query
+        }).execute()
+
+        # Handle a complex scenario where the tables column don't return immediately
+        if "jobComplete" in resp and not resp["jobComplete"]:
+            job_id = resp["jobReference"]["jobId"]
+
+            # First wait for the results to be calculated by BQ
+            state = "RUNNING"
+            while state == "RUNNING":
```

Comment:
couldn't that lead to an infinite loop ?

Reply:
no, bigquery isn't going to keep it in "running" forever